packages/flutter/lib/src/rendering/object.dart

```diff
@@ -361,6 +364,282 @@ typedef void RenderingExceptionHandler(RenderObject source, String method, dynam
 /// information will be printed to the console instead.
 RenderingExceptionHandler debugRenderingExceptionHandler;
 
+class _SemanticsGeometry {
+  _SemanticsGeometry() : transform = new Matrix4.identity();
+  _SemanticsGeometry.withClipFrom(_SemanticsGeometry other) {
+    clipRect = other?.clipRect;
+    transform = new Matrix4.identity();
+  }
+  _SemanticsGeometry.copy(_SemanticsGeometry other) {
+    if (other != null) {
+      clipRect = other.clipRect;
+      transform = new Matrix4.copy(other.transform);
+    } else {
+      transform = new Matrix4.identity();
+    }
+  }
+  Rect clipRect;
+  Rect _intersectClipRect(Rect other) {
+    if (clipRect == null)
+      return other;
+    if (other == null)
+      return clipRect;
+    return clipRect.intersect(other);
+  }
+  Matrix4 transform;
+  void applyAncestorChain(List<RenderObject> ancestorChain) {
+    for (int index = ancestorChain.length-1; index > 0; index -= 1) {
+      RenderObject parent = ancestorChain[index];
+      RenderObject child = ancestorChain[index-1];
+      clipRect = _intersectClipRect(parent.describeApproximatePaintClip(child));
+      if (clipRect != null) {
+        if (clipRect.isEmpty) {
+          clipRect = Rect.zero;
+        } else {
+          Matrix4 clipTransform = new Matrix4.identity();
+          parent.applyPaintTransform(child, clipTransform);
+          clipRect = MatrixUtils.transformRect(clipRect, clipTransform);
+        }
+      }
+      parent.applyPaintTransform(child, transform);
+    }
+  }
+  void updateSemanticsNode({ RenderObject rendering, SemanticsNode semantics, SemanticsNode parentSemantics }) {
```

Comment:
I wonder if we can extract a SemanticsProvider interface from RenderObject that we can use in these classes instead of RenderObject.  That might let us move this code out of object.dart into semantics.dart.